lib/pure/decimal.nim

```diff
@@ -0,0 +1,1040 @@
+      result[place] = result[place - 1]
+    result[0] = 0.byte
+
+
+proc shiftDecimalsRightTransient(values: TempSignificandArray, shiftNeeded: int16): TempSignificandArray =
+  for index in 0 ..< tempSignificandSize:
+    result[index] = values[index]
+  for _ in 0 ..< shiftNeeded:
+    for index in 1 ..< tempSignificandSize:
+      let place = tempSignificandSize - index
+      result[place] = result[place - 1]
+    result[0] = 0.byte
+
+proc digitCount(significand: SignificandArray): int =
+  # get the number of digits, ignoring the leading zeroes;
+  # special case: all zeroes results returns a result of zero
+  result = 0
+  var nonZeroFound = false
+  for d in significand:
+    if d != 0.byte:
+      nonZeroFound = true
+    if nonZeroFound:
+      result += 1
+
+proc digitCount(significand: TempSignificandArray): int =
+  # get the number of digits, ignoring the leading zeroes;
+  # special case: all zeroes results returns a result of zero
+  result = 0
+  var nonZeroFound = false
+  for d in significand:
+    if d != 0.byte:
+      nonZeroFound = true
+    if nonZeroFound:
+      result += 1
+
+proc toTempSignificandArray(original: SignificandArray): TempSignificandArray =
+  for index in 0 ..< tempSignificandSize:
+    let sIndex = index - transientOffset
+    if sIndex >= 0:
+      result[index] = original[sIndex]
+    else:
+      result[index] = 0.byte
+
+when not defined(js):
+  proc greaterOrEqualToOneBillion(val: Quotient): bool =
+    # is the number in the four 32-bit uints greater than 1_000_000_000
+    if val.a != 0'u32:
+      result = true
+    elif val.b != 0'u32:
+      result = true
+    elif ((val.c.uint64 shl 32) or val.d.uint64) >= billion.uint64:
+      result = true
+    else:
+      result = false
+
+  proc leftHalf(value: uint64): uint32 =
+    # get the left (most significant) half of a 64-bit uint
+    result = (value shr 32).uint32
+
+  proc setLeftHalf(value: var uint64, newValue: uint64) =
+    # when defined(js):
+    #   value = (value and lowerOnes) shr 0
+    #   let temp = (newValue and lowerOnes) shl 32  # shift new value to left
+    #   value = (value or temp) shr 0               # OR into place
+    # else:
+    value = value and lowerOnes
+    let temp = (newValue and lowerOnes) shl 32  # shift new value to left
+    value = value or temp                       # OR into place
+
+  proc rightHalf(value: uint64): uint32 =
+    # get the right (least significant) half of a 64-bit uint
+    result = ((value and lowerOnes) shr 0).uint32
+
+  proc setRightHalf(value: var uint64, newValue: uint64) =
+    value = (value and upperOnes) shr 0 # wipe out the right
+    value = (value or (newValue and lowerOnes)) shr 0
+
+proc divide(quotient: Quotient, divisor: uint32): (Quotient, uint32) =
+  when defined(js):
+    # TODO: write a version that only uses 32 bit uints
+    # For JS, this is needed for bitops
+    raise newException(Exception, "js support not written yet")
```

Comment:
```suggestion
    {.error: "js support not written yet".}
```
Ditto for other cases. Alternatively `{.fatal: "js support not written yet".}`, I'm not sure which one is more appropriate here, but this looks temporary anyway.